### PR TITLE
Allow with_request_url test helper can modify request.env to support Devise routes constraint

### PR DESCRIPTION
**Pull request by elct9620**
### What are you trying to accomplish?

The [devise](https://github.com/heartcombo/devise) provides the `authenticate` constraint but does not work correctly when using the `recognize_path` method.

But we can use `recognize_path_with_request` with the test helper created request object to make it work.

```ruby
before do
    # mock warden
    warden = instance_double(Warden::Manager)
    allow(warden).to receive(:send).with(:authenticate!, { scope: :admin_user }).and_return(true)
    request.env['warden'] = warden

    # render the component with the request
    with_request_url '/admin' do
      render_inline(component)
    end
end
```

### What approach did you choose and why?

The different between `Rails.application.routes.recognize_path` and `Rails.application.routes.recognize_path_with_request` is do we have provide the "request" object.

In our test helper, we already create a request object and we didn't need to use `Rack::MockRequest` created by Rails.

### Anything you want to highlight for special attention from reviewers?

It may have some side-effect if we didn't reset the request object between test cases.

### CODE REVIEW

File: lib/view_component/test_helpers.rb [hunk 1]
```diff
@@ -200,7 +200,7 @@ def with_request_url(path)
 
       path, query = path.split("?", 2)
       request.path_info = path
-      request.path_parameters = Rails.application.routes.recognize_path(path)
+      request.path_parameters = Rails.application.routes.recognize_path_with_request(request, path, {})
```
[the remark this answers is not in the document]
Comment (elct9620) [reply lib/view_component/test_helpers.rb:203]: Sure, let me work on it.

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -30,6 +30,10 @@ nav_order: 5
 
     *Anton Prins*
 
+* `with_request_url` test helper supports Devise router constraint.
```
Comment (joelhawksley) [inline docs/CHANGELOG.md:33]:
```suggestion
* `with_request_url` test helper supports router constraints (such as Devise).
```

**Review by joelhawksley — APPROVED** [review]

**Comment by simonc** [comment]
Note to anyone reaching this PR because your specs are broken after upgrading to v2.74.

If your specs use `with_request_url` with a fake path, your spec will break. If for instance you have:

```ruby
with_request_url('/some/random/path') do
  render_inline(...)
end
```

Calling `current_page?('/some/random/path')` will return `false` because

```ruby
request.path #=> "//some/random/path"
```

To fix your specs, simply switch to a public route that your application knows about and your specs should pass again.

❤️

**Comment by DRBragg** [comment]
Similar to [my comment](https://github.com/ViewComponent/view_component/pull/1550#issuecomment-1282633310) on #1550 this change breaks view_component's compatibility with Rails <= 5.1 because it appears `recognize_path_with_request` isn't in `ActionDispatch::Routing::RouteSet` until 5.2

**Comment by elct9620** [comment]
Maybe we can give a fallback to `recognize_path` if Rails <= 5.1 🤔 

```ruby
if Rails::VERSION::MAJOR <= 5 && RAILS::VERSION::MINIOR <= 1
   # `request.env` may not fully fit the requirement here
   request.path_parameters = Rails.application.routes.recognize_path_with_request(request, path, request.env)
else
  # ...
end
```

**Pull request merged**